### DISCUSSION TITLE
$wire.someMethod(someParam).then(result => { ... }) always has NULL result.

### DISCUSSION THREAD

**Discussion by trip-somers** [livewire/livewire#5424] [7. Bugs (archived)]
```javascript
// Calling a Livewire method and doing something with its result
$wire.someMethod(someParam)
    .then(result => { ... })
```

The above snippet is from the Livewire 2.x documentation on this page -- https://laravel-livewire.com/docs/2.x/alpine-js#interacting-with-livewire-from-alpine -- but every time we attempt to use this, `result` is null. It seems like `.then()` is not awaiting `$wire.someMethod(someParam)`.

I have searched all over the normal places (google, stackoverflow, github issues, etc.) for other people experiencing this problem, and I have come up empty.

**Comment by joshhanley** [comment]
@trip-somers the result will be whatever you returned from the method, so if you don't have a return then it will be null.

Try returning something from the component class method and see if that works.

Hope this helps!

**Reply by trip-somers** [reply]
Yep, we tried that, too.

**Reply by trip-somers** [reply]
Hmmmm... but maybe we were expecting to still get a result when a validation error is encountered.

**Reply by trip-somers** [reply]
We've confirmed that it actually works as you described when we aren't intentionally trying to trigger an error. Our use-case was only closing a modal upon success, so we were testing that it stayed open when an error was present. We tried `console.log(result)` to see what we had to work with, but it was always null.

It makes sense now. Thank you for confirming the expected value of `result`.

**Reply by trip-somers** [reply]
We have a more complex case now.

We have a Save button that returns a redirect on success. The button gets disabled on first click, but obviously, we need to remove the 'disabled' attribute if there's an error. The catch here is that `result` is `null` when redirecting and `null` when there's an error. When we receive a success response (a redirect) the button is being re-enabled in the space between when the browser receives the redirect and when it handles it. In this window, duplicate submissions can occur.

How can we detect the difference between a successful redirect and an error response?

**Reply by joshhanley** [reply]
@trip-somers hmm I'm not sure you can distinguish the difference, but Livewire should auto handle the redirect anyway, so not sure why that would matter. Any chance you can give me some more context and background with exactly what you're trying to do/achieve with code examples? Maybe there is a different way to approach it.

**Reply by trip-somers** [reply]
The roundtrip involves validation, and if that fails, we need to re-enable the button. If it succeeds, we want to keep the button disabled so that a second submission won't happen.

```html
<div
    x-data="{
        storing: false,
        store() {
            this.storing = true;

            $wire.store().then(result => {
                if (! result) {
                    this.storing = false;
                }
            });
        }
    }"
>
```

```html
<button type="button" @click="store()" :disabled="storing" class="btn primary">
    Continue
</button>
```

The catch at this point, is that `result` is null for a redirect _and_ for a validation failure, so this code re-enables the button in the browser while the browser handles the redirect. The window is relatively narrow, but long enough, especially for mobile internet, that we need to prevent a second submission.

**Reply by joshhanley** [reply]
@trip-somers ah yep, yeah that makes sense. I think you might be able to make use of dispatching a browser event in the event of a redirect or validation failure and listen for that on the front end to control button state.
